### PR TITLE
BT enablement on fairseq - fairseq change

### PR DESCRIPTION
**Pull request by frank-wei**
Summary:
as titled and depends on D36057338
Fork the inference path inside the forward function. If loaded the checkpoint file and perform the inference, we will deploy BT. Otherwise, fairseq take the position.

In summary:
Accuracy: accuracy loss due to the fp16, the maximum diff is around 0.009. If we set it to fp32, there is no accuracy loss
Perf: the current fairseq has similar speed as vanilla version. After the enablement, the speedup is similar to standalone BT test.
With batch size=64
For V100, the speedup reaches to 1.23x
For A100, the speedup reaches to 1.38x

After enable nested tensor,
For V100, the speedup reaches to 2.46x

Reviewed By: mikekgfb

Differential Revision: D37082681

### CODE REVIEW

**Review by mikekgfb — APPROVED** [review]

File: fairseq/modules/transformer_layer.py [hunk 1]
```diff
@@ -187,44 +283,83 @@ def forward(
         # Note that we cannot use -inf here, because at some edge cases,
         # the attention weight (before softmax) for some padded element in query
         # will become -inf, which results in NaN in model parameters
-        if attn_mask is not None:
-            attn_mask = attn_mask.masked_fill(
-                attn_mask.to(torch.bool), -1e8 if x.dtype == torch.float32 else -1e4
+
+        if self.training:
+            self.ever_training = True
```
Comment (dianaml0) [inline fairseq/modules/transformer_layer.py:288]: Why have a variable that is the same as self.training?

File: tests/test_sequence_generator.py [hunk 2]
```diff
@@ -113,7 +140,9 @@ def _test_save_and_load(self, scripted_module):
 JIT_MSG = "Targeting OSS scriptability for the 1.6 release"
 
 
-@unittest.skipIf(torch.__version__ < "1.6.0", JIT_MSG)
+@unittest.skipIf(
```
Comment (dianaml0) [inline tests/test_sequence_generator.py:143]: You can add the necessary pytorch version to the CI testing to make sure this test works and passes.

File: fairseq/modules/transformer_layer.py (hunk 1 above)
Comment (frank-wei) [reply fairseq/modules/transformer_layer.py:288]: That is for the scenario where loading ckpt+continue training. We'd like to prevent this case to go fast path.

File: tests/test_sequence_generator.py (hunk 2 above)
Comment (frank-wei) [reply tests/test_sequence_generator.py:143]: I can add one nightly version like 1.13.0.dev20220613 as one more test.

Comment (dianaml0) [reply tests/test_sequence_generator.py:143]: That sounds good to me

**Comment by frank-wei** [comment]
@dianaml0  The flow creation could be separate from this PR so I landed it first. 
I am not allowed to create any flow or yaml file as I met such error:
 new_flow -> new_flow (refusing to allow a Personal Access Token to create or update workflow `.github/workflows/build_1_13.yml` without `workflow` scope)
 
 Could you help to create a flow with new yaml file? It should be 99% the same with your current `build.yml` but changed the line 31 to:
```
 run: pip3 install --pre torch==1.13.0.dev20220613 -f https://download.pytorch.org/whl/nightly/torch_nightly.html 
```
